File: tutorial/T18-Event.cpp
/***************************
@Author: Chunel
@Contact: chunel@foxmail.com
@File: T18-Event.cpp
@Time: 2023/1/20 17:55
@Desc: 本例子主要展示触发Event功能
***************************/

#include "MyGNode/MyNode1.h"
#include "MyGNode/MyWriteParamNode.h"
#include "MyGNode/MyEventNode.h"
#include "MyGEvent/MyPrintEvent.h"

using namespace CGraph;

void tutorial_event() {
    GPipelinePtr pipeline = GPipelineFactory::create();
    GElementPtr a, b, c, d = nullptr;

    CStatus status = pipeline->registerGElement<MyWriteParamNode>(&a, {}, "nodeA");
    status += pipeline->registerGElement<MyEventNode>(&b, {a}, "nodeB");
    status += pipeline->registerGElement<MyNode1>(&c, {b}, "nodeC");
    status += pipeline->registerGElement<MyEventNode>(&d, {b}, "nodeD");

    // 在pipeline中添加一个事件信息，通过 notify("my-print-event")触发
    pipeline->addGEvent<MyPrintEvent>("my-print-event");
    pipeline->process();

    GPipelineFactory::clear();
}


int main() {
    tutorial_event();
    return 0;
}

File: python/tutorial/T18-Event.py
"""
@Author: Chunel
@Contact: chunel@foxmail.com
@File: T18-Event
@Time: 2025/3/22 00:21
@Desc: 
"""

from pycgraph import GPipeline

from MyGNode.MyNode1 import MyNode1
from MyGNode.MyEventNode import MyEventNode
from MyGNode.MyWriteParamNode import MyWriteParamNode
from MyGEvent.MyPrintEvent import MyPrintEvent


def tutorial_event():
    pipeline = GPipeline()
    a, b, c, d = MyWriteParamNode(), MyEventNode(), MyNode1(), MyEventNode()

    pipeline.registerGElement(a, set(), "nodeA")
    pipeline.registerGElement(b, {a}, "eventNodeB")
    pipeline.registerGElement(c, {b}, "nodeC")
    pipeline.registerGElement(d, {c}, "eventNodeD")

    pipeline.addGEvent(MyPrintEvent(), "event1")

    pipeline.process()


if __name__ == '__main__':
    tutorial_event()
